### DISCUSSION TITLE
Embeded Error Messages

### DISCUSSION THREAD

**Discussion by chanchantang** [CSSS/wall_e#665] [Q&A] [answered]
I am trying to embed an error message in `sfu.py`, but I am getting an error trying to do so.
Here is the code in question:
```
e_obj = await embed(
    self.logger, interaction=interaction, title='SFU Courses Error',
    description=(
        f'Couldn\'t find anything for `{", ".join(departments)}, level {level}, term {term}, year {year}`'
        f'\n Maybe the course doesn\'t exist? Or isn\'t offered right now.'
    ),
    colour=WallEColour.ERROR,
)
if e_obj:
    msg = await interaction.followup.send(embed=e_obj)
```
The error I get is
![image](https://github.com/user-attachments/assets/dbb09c97-ee25-4237-b54c-2e54ab92da7a)
If I had to guess, it would have to do something with this being a slash command but idk
Here is the [entire code](https://github.com/chanchantang/wall_e/blob/2182863f957bc06e304d2e8e1835d9552c83f648/wall_e/extensions/sfu.py#L513C15-L513C22 ) if needed

**Comment by modernNeo** [comment]
if I am not misremembering, you use `interaction.followup.send(embed=e_obj)` for an interaction where you have deferred it via `await interaction.response.defer()`

If you haven't deferred, you would only need `interaction.response.send_message(embed=e_obj)`

and just so you know, you only need to defer an interaction if you anticipate that the command you are running will take longer than 3 seconds to compute and respond to the command.

**Reply by chanchantang** [reply]
Hmm, I'm confused on how to use defer. 
For example, in the `bans()` function, it uses defer in the beginning, then an `await BanRecord` query, then parsing it. I'm guessing this could take more than 3 seconds, so we use defer and followup.
But for the `ban()` function, it uses `interaction.response.defer()` in the beginning as well, then right after a `interaction.followup.send` if the purge window days are outside the range. Is it doing this because performing `wall_e_ban()` might take more than 3 seconds? If so, why doesn't the ban `intercept()` function need to use it? 

But `interaction.response.send_message(embed=e_obj)` works perfectly, thanks a lot!

**Reply by modernNeo** [reply]
>  For example, in the bans() function, it uses defer in the beginning, then an await BanRecord query, then parsing it. I'm guessing this could take more than 3 seconds, so we use defer and followup.

Correct.

>  But for the ban() function, it uses interaction.response.defer() in the beginning as well, then right after a interaction.followup.send if the purge window days are outside the range. Is it doing this because performing wall_e_ban() might take more than 3 seconds? If so, why doesn't the ban intercept() function need to use it?

Can you point me to where the ban function uses `interaction.response.defer` right after the `await interaction.followup.send`?

**Reply by chanchantang** [reply]
Oh my apologies I miswrote, I meant it uses followup right after defer, ie `defer()` -> `followup()`. 

My main question was meant to ask how `ban()` uses `defer()` -> `followup()` for error -> `wall_e_ban()`,
but then within `wall_e_ban()` it does not use followup in its function

Also, if the main concern was the calling of `wall_e_ban()`, the function `intercept()`, which uses `wall_e_ban()`, does no use defer nor followup

Sorry if this does not make sense

**Reply by modernNeo** [reply, marked as answer]
> Oh my apologies I miswrote, I meant it uses followup right after defer, ie defer() -> followup().

Are you referring to this bit here?
https://github.com/CSSS/wall_e/blob/c0e6d39cb16abaeb096e3f36d033ccd3eece6c53/wall_e/extensions/ban.py#L404-L418

> My main question was meant to ask how ban() uses defer() -> followup() for error -> wall_e_ban(),
but then within wall_e_ban() it does not use followup in its function

So, there are a few different things at play there.
given that `ban` is a slash command, that means that the command **has** to respond to the invoking message somehow [to use proper discord.py terminology, the interaction has to be responded to], this can be done either via
 * `interaction.response.send_message` if you know the code will take <= 3 seconds to do its thing
 *  or via `interaction.response.defer`/`interaction.followup.send` if the response to the invoking message will take longer than 3 seconds.

**However**, the discord.py additionally allows us to send message in different contexts:
1. for example, we are given to option to send a stand-alone message just in the channel where the invoking message was called from **but not as a response to the invoking message**:
https://github.com/CSSS/wall_e/blob/c0e6d39cb16abaeb096e3f36d033ccd3eece6c53/wall_e/extensions/ban.py#L407-L410
2. in the `wall_e_ban` function, if the interaction object is not specified, then the message are just sent to the mod channel via this function: https://github.com/CSSS/wall_e/blob/c0e6d39cb16abaeb096e3f36d033ccd3eece6c53/wall_e/extensions/ban.py#L251-L253
3. or we can aways send a DM directly to the user directly like this [assuming they haven't disabled DMs]: https://github.com/CSSS/wall_e/blob/c0e6d39cb16abaeb096e3f36d033ccd3eece6c53/wall_e/extensions/ban.py#L350

But to more directly answer your question,

if you look back to https://github.com/CSSS/wall_e/blob/c0e6d39cb16abaeb096e3f36d033ccd3eece6c53/wall_e/extensions/ban.py#L251-L252

you will notice that if the interaction parameter is not `None`, the function used is not `interaction.followup.send`, but `interaction.edit_original_response`.

What is going on is a user invoked the ban command, thereby creating an interaction that WALL_E has to respond to in some way. So the `ban` command first tells the discord API for that command that it needs to wait longer than 3 seconds for the reply via `interaction.response.defer` before eventually responding via `interaction.followup.send`. but even once the command has responded, while you cannot re-repond even if you have new information to relay to the user, you can either just send another non-interaction message tagging the original user or you can modify the response [edit the message that the bot responded back with] by using `interaction.edit_original_response` to include the new info.

Hope that clears it up.

**Reply by chanchantang** [reply]
Oooh I see, I think I get it now. Thanks for explaining it so thoroughly!
